// @creditkarma/thrift-client/dist/main/connections/pool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const GenericPool = require("generic-pool");
const Connection_1 = require("./Connection");
const defaultOptions = {
    min: 0,
    max: 10,
    evictionRunIntervalMillis: 10000,
    idleTimeoutMillis: 10000,
    acquireTimeoutMillis: 1000,
    testOnBorrow: true,
};
exports.createPool = (config, logger, options) => {
    const resolvedOptions = Object.assign(defaultOptions, options);
    const factory = {
        create: async () => {
            logger(['debug', 'createPool'], 'Creating new client connection');
            return await Connection_1.createConnection(config);
        },
        destroy: async (connection) => {
            logger(['debug', 'createPool'], 'Destroying client connection');
            return connection.destroy().then(() => undefined);
        },
        validate: async (connection) => {
            return connection.hasSession();
        },
    };
    return GenericPool.createPool(factory, resolvedOptions);
};
